### DISCUSSION TITLE
How to propagate a PropertyAttribute on a class to its methods in the XML output?

### DISCUSSION THREAD

**Discussion by marco-eckstein** [nunit/nunit#4713] [How to]
I have annotated a test class with an NUnit.Framework.PropertyAttribute. This results in the XML test results file having this attribute as property child element of the element that represents this test class.

Now I also want this attribute as property child element of the elements that represent the test methods, but I do not want to annotate the test methods with this attribute explicitly. What are the best ways to do this?

It may be possible by writing an XSLT stylesheet and [passing its location to the command line runner](https://stackoverflow.com/a/2222262/443836). But since I have no experience with XSLT, I may prefer another approach.

Also posted[ at Stack Overflow](https://stackoverflow.com/q/78488503/443836).
I am using NUnit 3.

**Comment by mikkelbu** [comment]
Hi @marco-eckstein

Instead of using XSLT you can make your own variant of the property attribute that will also apply the property to the tests - e.g. like this
```csharp
    public class MyAttribute: PropertyAttribute, IApplyToTest
    {
        private readonly string propertyName;
        private readonly string propertyValue;

        public MyAttribute(string propertyName, string propertyValue) : base(propertyName, propertyValue)
        {
            this.propertyName = propertyName;
            this.propertyValue = propertyValue;
        }

        public override void ApplyToTest(Test test)
        {
            base.ApplyToTest(test);

            if (test.HasChildren)
            {
                foreach (var child in test.Tests)
                {
                    if (child is TestMethod testMethod)
                    {
                        testMethod.Properties.Add(propertyName, propertyValue);
                    }
                }
            }
        }
    }
```
and then you can use it like this (note that I had to explicitly add `[TestFixture]` for the attribute to discover the child tests)
```csharp
    [My("MyProperty", "MyValue")]
    [TestFixture]
    public class Tests
    {
        [Test]
        public void Test1()
        {
            Assert.Pass();
        }
    }
```
